controller/rest/vulasset.go

```diff
@@ -466,14 +545,155 @@ func combineQueryFilter(r *http.Request) (*db.VulQueryFilter, error) {
 
 	return queryFilter, nil
 }
+
+func createAssetSession(w http.ResponseWriter, r *http.Request) {
+	log.WithFields(log.Fields{"URL": r.URL.String()}).Debug("")
+	defer r.Body.Close()
+
+	queryStat := &api.RESTAssetQueryStats{
+		PerfStats: make([]string, 0),
+	}
+
+	acc, login := getAccessControl(w, r, access.AccessOPRead)
+	if acc == nil {
+		return
+	} else if !acc.HasRequiredPermissions() {
+		restRespAccessDenied(w, login)
+		return
+	}
+
+	queryFilter, err := db.GetAssetQuery(r)
+	if err != nil {
+		restRespErrorMessage(w, http.StatusBadRequest, api.RESTErrInvalidRequest, err.Error())
+		return
+	}
+
+	if queryFilter.Filters.Type != "image" {
+		restRespErrorMessage(w, http.StatusBadRequest, api.RESTErrInvalidRequest, "invalid asset type")
+		return
+	}
+
+	// create a query session
+	queryFilter.QueryToken = utils.GetRandomID(6, "") // do not change the length
+
+	queryFilterb, _ := json.Marshal(queryFilter)
+	err = createQueryStat(login, 1, queryFilter.QueryToken, string(queryFilterb))
+	if err != nil {
+		log.WithFields(log.Fields{"err": err}).Error("createQueryStat error")
+		restRespErrorMessage(w, http.StatusInternalServerError, api.RESTErrInvalidRequest, err.Error())
+		return
+	}
+
+	// notify other controllers via Consul
+	err = createRequestInConsul(acc, login, 1, queryFilter.QueryToken, nil, queryFilter.Filters)
+	if err != nil {
+		log.WithFields(log.Fields{"err": err}).Error("Write create session request to consul error")
+	}
+
+	start := time.Now()
+	allowed := getAllAllowedResourceId(acc)
+	elapsed := time.Since(start)
+
+	queryStat.PerfStats = append(queryStat.PerfStats, fmt.Sprintf("1/4, get allowed resources, asset_count=%d, took=%v", allowed[db.AssetWorkload].Cardinality(), elapsed))
+
+	assetCount, top5, err := db.CreateImageAssetSession(allowed, queryFilter)
+	if err != nil {
+		restRespErrorMessage(w, http.StatusInternalServerError, api.RESTErrInvalidRequest, err.Error())
+		return
+	}
+
+	// get statistics and return to caller
+	queryStat.TotalRecordCount = assetCount
+	queryStat.QueryToken = queryFilter.QueryToken
+	queryStat.Summary = &api.AssetSessionSummary{}
+	queryStat.Summary.TopImages = top5
+
+	log.WithFields(log.Fields{"PerfStats": strings.Join(queryStat.PerfStats, ";"), "querytoken": queryStat.QueryToken}).Debug("createAssetSession")
+
+	if queryFilter.Debug == 0 {
+		queryStat.PerfStats = nil
+	}
+
+	records, err := db.GetExceededSessions(login.fullname, login.id, login.loginType)
+	if err == nil {
+		for _, token := range records {
+			clusHelper.DeleteQuerySessionRequest(token)
+		}
+		log.WithFields(log.Fields{"records": records}).Debug("Delete exceeded sessions")
+	}
+
+	restRespSuccess(w, r, queryStat, acc, login, nil, "Create asset session")
+
+	// populate the result to file-based database
+	go func() {
+		err = db.DupAssetSessionTableToFile(queryFilter.QueryToken)
+		if err != nil {
+			log.WithFields(log.Fields{"err": err}).Error("PopulateSessionToFile error")
+		}
+	}()
```

Comment:
Question: Is there a reason why this has to be deferred to a go routine?

Reply:
for better performance as this is a time consuming task and it's inside the http request flow.